Added vector datatype support in Oracle dialect

suraj-ora-2020 opened the pull request:
<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->
This PR adds support for the VECTOR datatype in the Oracle dialect of SQLAlchemy, which corresponds to the vector data type introduced in Oracle Database 23ai. This new VECTOR type allows for efficient storage and retrieval of high-dimensional vector data, enabling SQLAlchemy users to leverage vector-based queries. By treating VECTOR as a "first-class" type, this integration makes it easier to work with Oracle’s vector-based capabilities directly in Python.

Fixes: #12317 

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

suraj-ora-2020 commented:
@CaselIT I have created a new PR. This have only changes related to vector datatype. For fetch_type i will create a new PR.

CaselIT commented:
Thank you, we will take a look soon!

zzzeek commented on lib/sqlalchemy/dialects/oracle/oracledb.py:
```diff
@@ -591,6 +591,228 @@ def _remove_clob(inputsizes, cursor, statement, parameters, context):
+Top-Level Keys
+==============
+
+These keys are specified directly under the ``oracle_vector`` dictionary.
```
why not use a dataclass for this?  This is a highly complex configuration, I'd use modern python constructs

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -1542,6 +1557,45 @@ def visit_create_index(self, create, **kw):
+            if using == "HNSW":
+                parts.append("ORGANIZATION INMEMORY NEIGHBOR GRAPH")
+            elif using == "IVF":
+                parts.append("ORGANIZATION NEIGHBOR PARTITIONS")
```
all of this validation stuff I think should be handled using dataclasses + enums up front for this config since this is very complex config

zzzeek reviewed (changes requested):
I would recommend using dataclasses + enums for this index configuration since all the dictionary  / hardcoded keywords / checks are too verbose for modern Python and place burden on the developer to copy and paste from docs, rather than allowing their IDE to help with most of it

I think we will need some more passes through the tests as well to clean things up

@CaselIT any comments?

suraj-ora-2020 commented:
@CaselIT I've incorporated the suggested feedback and updated the code accordingly.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 5505f00da43cccac86d80fbf1b9cd382857bcaa4 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 5505f00da43cccac86d80fbf1b9cd382857bcaa4: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

zzzeek commented:
just needs pep8.   i will need to review more closely since this is a big change

zzzeek commented:
sorry we are moving slowly on this, can you install pre-commit and re-commit so pep8 fixes are implemented?

suraj-ora-2020 commented:
@zzzeek made the changes. I hope pep8 issues are fixed.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 53d1fc416697169e380cb9e6128d0418c40bde86 of this pull request into gerrit so we can run tests and reviews and stuff

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -744,14 +744,208 @@
+When using Oracle VECTOR indexes, the configuration parameters are divided based on index type
+**HNSW** and **IVF**. This structure applies to both HNSW and IVF VECTOR indexes.
+
+Comman Attributes
```
typo

Also I think 99% of this section should be in the docstring for `VectorIndexConfig` and not here in the main docstring

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -850,9 +1045,91 @@
 }
 
 
+class VectorIndexType(Enum):
```
this needs a docstring

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -850,9 +1045,91 @@
+    IVF = "IVF"
+
+
+class VectorDistanceType(Enum):
```
needs a docstring, docstrings for elements as well

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -850,9 +1045,91 @@
+
+
+@dataclass
+class VectorIndexConfig:
```
this needs a docstring also, but it also has to list out these params.  since there is no `__init__` method, let's try documenting the params here.  Additionally, as they seem to be specific to various `VectorIndexType` combinations, that should be here as well:

```py
@dataclass
class VectorIndexConfig:
     """Defines the configuration for < accurate terminology here >

      :param index_type: the <XYZ> - applies to :attr:`.VectorIndexType.HNSW`, <others>
      :param distance: <QPR> - applies to :attr:`.VectorIndexType.FOO`, :attr:`.VectorIndexType.BAR`
      # ... etc.

```

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -850,9 +1045,91 @@
+                )
+
+
+def build_vector_index_config(vector_index_config: VectorIndexConfig) -> str:
```
this should be an underscored method on OracleCompiler

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -744,14 +744,208 @@
+
+* ``distance``:
+    - Specifies the metric for calculating distance between VECTORS.
+    - **Valid Values**: Enum values from `VectorDistanceType` (`EUCLIDEAN`, `COSINE`, `DOT`, `MANHATTAN`).
```
so throughout all these docs, the classes have to be referred to like this, using sphinx paramlinks directives:

    - **Valid Values**: Enum values from :class:`.VectorDistanceType`

    - Specifies the indexing method. For HNSW, this must be :attr:`.VectorIndexType.HNSW`


Then these classes need to be added to https://github.com/sqlalchemy/sqlalchemy/blob/main/doc/build/dialects/oracle.rst using `.. autoclass:`   - `VECTOR` should be added to the types section, then the index constructs will be in a new section `Oracle DDL Constructs` which will have a dotted line underscore `------`.   then they will be linked from these docs

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -744,14 +744,208 @@
+* ``parallel``:
+    - Specifies degree of parallelism
+
+HNSW Parameters
```
So the descriptions of these parameters will be in the docstring for `VectorIndexConfig`, they dont need to be fully expanded upon here

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -850,9 +1045,91 @@
 
+class VectorIndexType(Enum):
+    HNSW = "HNSW"
+    IVF = "IVF"
```
the elements can have docstrings also

```py
class VectorIndexType(Enum):
    HNSW = "HNSW"
    """the HNSW index type. 

    Configuration parameters for this type include :paramref:`.VectorIndexConfig.distance`, :paramref:`.VectorIndexConfig.foobar`
   
   """

   IVF = "IVF"
   """The IVF index type

   etc

   """

```

zzzeek reviewed (changes requested):
I'm proposing moving the parameter-specific documentation to be in `VectorIndexConfig` and related classes, so that it will be much more cross-linked.

As this will greatly increase the line length of these classes, I propose that all of:

`VECTOR`, `VectorIndexConfig`, `VectorDistanceType`, `VectorIndexType` be moved to a new source file lib/sqlalchemy/dialects/oracle/vector.py since this is a big new subject of its own

suraj-ora-2020 commented:
> it looks like a class called `VectorStorageFormat` has gotten lost

yes this line was missing from `.types import VectorStorageFormat` . And thanks for reviewing I will update the PR as per your suggestion.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision a0b3fa8fc993d2f70a089a649edb1cf18b12a84a of this pull request into gerrit so we can run tests and reviews and stuff

suraj-ora-2020 committed:
Added vector datatype support in Oracle dialect

Added new datatype :class:`_oracle.VECTOR` and accompanying DDL and DQL
support to fully support this type for Oracle Database. This change
includes the base :class:`_oracle.VECTOR` type that adds new type-specific
methods ``l2_distance``, ``cosine_distance``, ``inner_product`` as well as
new parameters ``oracle_vector`` for the :class:`.Index` construct,
allowing vector indexes to be configured, and ``oracle_fetch_approximate``
for the :meth:`.Select.fetch` clause.  Pull request courtesy Suraj Shaw.

Fixes: #12317
Fixes: #12341
Closes: #12321
Pull-request: https://github.com/sqlalchemy/sqlalchemy/pull/12321
Pull-request-sha: a0b3fa8fc993d2f70a089a649edb1cf18b12a84a

Change-Id: I6f3af4623ce439d0820c14582cd129df293f0ba8

Added vector datatype support in Oracle dialect #12321

Added vector datatype support in Oracle dialect#12321

Added vector datatype support in Oracle dialect#12321

updated to static link

some doc modification

sqla-tester commented:
Patchset a0b3fa8fc993d2f70a089a649edb1cf18b12a84a added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

* lib/sqlalchemy/dialects/oracle/base.py (line 1414): I just realized this method is copied from base compiler.py.  All it does is add APPROX to the final result.   Can we just use the super() implementation and simply inject APPROX right after the ROWS indicator

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

* lib/sqlalchemy/dialects/oracle/base.py (line 1414): Done

sqla-tester commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -744,11 +744,140 @@
+
+    vector_data_8 = [1, 2, 3]
+    statement  = insert(t1)
+    with engine.connect() as conn:
```
**Federico Caselli** (CaselIT) wrote:

these 3 sections above would be nicer if word wrapped to ~ 80 chars.

it's just a preference, the text is otherwise fine

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+
+    :param distance: Enum values from :class:`.VectorDistanceType`
+     specifies the metric for calculating distance between VECTORS.
+
```
**Federico Caselli** (CaselIT) wrote:

I guess reflecting these is not easy, but we could just test that it doesn't break anything?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+
+    :param parallel: integer. Specifies degree of parallelism.
+
+    :param hnsw_neighbors: interger. Should be in the range 0 to
```
**Federico Caselli** (CaselIT) wrote:

values -> value

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+     2048. Specifies the number of nearest neighbors considered
+     during the search The attribute :attr:`VectorIndexConfig.hnsw_neighbors`
+     is HNSW index specific.
+
```
**Federico Caselli** (CaselIT) wrote:

values -> value

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+     0 to 10,000,000. Specifies the number of partitions used to
+     divide the dataset. The attribute
+     :attr:`VectorIndexConfig.ivf_neighbour_partitions` is IVF index
+     specific.
```
**Federico Caselli** (CaselIT) wrote:

missing period after search

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+    :param ivf sample_per_partition: integer. Should be between 1
+     and ``num_vectors / neighbor partitions``. Specifies the
+     number of samples used per partition. The attribute
+     :attr:`VectorIndexConfig.ivf sample_per_partition` is IVF index
```
**Federico Caselli** (CaselIT) wrote:

missing space after period

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+    """
+
+    index_type: VectorIndexType = VectorIndexType.HNSW
+    distance: Optional[VectorDistanceType] = None
```
**Federico Caselli** (CaselIT) wrote:

missing _ in the param name

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+    distance: Optional[VectorDistanceType] = None
+    accuracy: Optional[int] = None
+    hnsw_neighbors: Optional[int] = None
+    hnsw_efconstruction: Optional[int] = None
```
**Federico Caselli** (CaselIT) wrote:

missing _ in the param name

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+    ivf_sample_per_partition: Optional[int] = None
+    ivf_min_vectors_per_partition: Optional[int] = None
+    parallel: Optional[int] = None
+
```
**Federico Caselli** (CaselIT) wrote:

I'm guessing this is missing
>  The attribute :attr:`VectorIndexConfig.ivf_min_vectors_per_partition` is IVF index specific.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,243 @@
+         should be an integer value.
+
+        :param storage_format: VectorStorageFormat. The VECTOR storage
+         type format. This may be Enum values form
```
**Federico Caselli** (CaselIT) wrote:

do we need to make some changes to support reflection? if it already works a test that leverages it would be nice to have

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

i think we definitely would need additional logic to support reflection since the type accepts arguments, and one of them is quite complex.  The VECTOR would have to be added to the map of type lookups to even come back from reflection so yes, this is not done at all

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on test/dialect/oracle/test_types.py:
```diff
@@ -951,6 +958,194 @@ def test_longstring(self, metadata, connection):
         finally:
             exec_sql(connection, "DROP TABLE Z_TEST")
 
+    @testing.fails_if("oracle<=23.4")
```
**Federico Caselli** (CaselIT) wrote:

would only_on be better here (also in the others)? it makes little sense to try it on older versions since we know vector is not there

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

some docs and test suggestions.

also reflection is not tested

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

* lib/sqlalchemy/dialects/oracle/base.py (line 740): 8bit unsigned is called binary? it may be worth mentioning it in parenthesis?
something like `8-bit unsigned integers (binary)`
* lib/sqlalchemy/dialects/oracle/base.py (line 1433): no need of `r` here
* lib/sqlalchemy/dialects/oracle/vector.py (line 247): since this is used in a bind processor it may make sense to cache this mapping, either in the VECTOR class or at the module level
* lib/sqlalchemy/sql/selectable.py (line 3894): @mike I guess this kinda overlaps with Select.ext, but we don't have it available for 2.0.

I guess moving forward simple boolean modifiers of existing select features are fine as dialect kwargs, but more complex things would need to be extension, otherwise it could seem a bit arbitrary what is a dialect kwargs and what's an extension

CaselIT commented:
@suraj-ora-2020 the change was changed on gerrit, so either me or mike will address the comments. Thanks for the contribution

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

@suraj -

I've modified this PR in gerrit beyond what you have, however, @caselit points out that we need to have some reflection tests to at least ensure if someone reflects a database with a VECTOR datatype, it comes back, or with a VECTOR index, it at least does not crash.  ideally the arguments for these two objects would be fully reflectable but I will say that can be optional here.  but he's right that we need reflection round trip tests.

Can you pull in the current state of this code by git fetching from it:

    git fetch https://gerrit.sqlalchemy.org/sqlalchemy/sqlalchemy refs/changes/38/5838/7 && git checkout FETCH_HEAD
    
then push the current state of the patch to your PR in github, then implement some reflection round trip tests?   that's what we'd need to continue here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

* lib/sqlalchemy/sql/selectable.py (line 3894): if the extension makes use of a ClauseElement or any kinds of parameters then it has to be ext().   otherwise, if it's a string or boolean that is acceptable as a part of the static cache key then it can be dialectkwargs.

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

* lib/sqlalchemy/sql/selectable.py (line 3894): makes sense to me

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 0a471c306013205759ebfc714bba52fe329d734e of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 0a471c306013205759ebfc714bba52fe329d734e added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

suraj-ora-2020 commented:
@CaselIT @zzzeek  Thanks for the feedback! I've made all the requested changes. Let me know if anything else needs updating.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 22c34f67e7b53ee12385afd40c504bfc820c3c42 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,265 @@
+
+
+@dataclass
+class VectorIndexConfig:
```
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,265 @@
+                    f"{field} must be an integer if"
+                    f"provided, got {type(value).__name__}"
+                )
+
```
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -730,11 +730,176 @@
+
+Oracle Database 23ai introduced a new VECTOR datatype for artificial intelligence
+and machine learning search operations. The VECTOR datatype is a homogeneous array
+of 8-bit signed integers, 8-bit unsigned integers (binary), 32-bit floating-point numbers,
```
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -730,11 +730,176 @@
+    from sqlalchemy import insert, select
+
+    with engine.begin() as conn:
+        conn.execute(
```
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -1234,6 +1414,29 @@ def _get_limit_or_fetch(self, select):
+                use_literal_execute_for_simple_int
+            ),
+            **kw,
+        )
```
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

suraj-ora-2020 commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -730,11 +730,176 @@
+
+.. seealso::
+
+    `CREATE VECTOR INDEX <https://www.oracle.com/pls/topic/lookup?ctx=dblatest&id=GUID-B396C369-54BB-4098-A0DD-7C54B3A0D66F>`_ - in the Oracle documentation
```
Yes, its a redirecting lookup link so that it is not required to update the link with each new db release. With the earlier one it was required to update the link with every db release.

suraj-ora-2020 commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,265 @@
+
+    :param ivf_min_vectors_per_partition: integer. From 0 (no trimming)
+     to the total number of vectors (results in 1 partition). Specifies
+     the minimum number of vectors per partition. The attribute
```
Which one is right
> :attr:`.VectorIndexConfig.ivf_min_vectors_per_partition`

> :attr:`VectorIndexConfig.ivf_min_vectors_per_partition`

In first there is period before VectorIndexConfig. If this one is correct then other places also needs to be modified.

zzzeek replied:
the period allows sphinx to do a search to get the correct namespace for the object, which in this case would be somehting like `sqlalchemy.dialects.oracle.VectorIndexConfig`.  that is, the period tells sphinx to substitute the full module path.   if the parent document of this doc already sets up `:module:` to be `sqlalchemy.dialects.oracle`, then that's the default prefix in any case, but I tend not to rely on that in most cases since sphinx can be very unreliable.

background is at https://www.sphinx-doc.org/en/master/usage/domains/python.html#target-resolution

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision a72a18a45c85ae7fa50a34e97ac642e16b463b54 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset a72a18a45c85ae7fa50a34e97ac642e16b463b54 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,266 @@
+        """
+
+        def process(value):
+            if isinstance(value, array.array):
```
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented on lib/sqlalchemy/dialects/oracle/vector.py:
```diff
@@ -0,0 +1,266 @@
+    def _array_typecode(self, typecode):
+        """
+        Map storage format to array typecode.
+        """
```
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

zzzeek commented on lib/sqlalchemy/dialects/oracle/base.py:
```diff
@@ -730,11 +730,176 @@
+.. seealso::
+
+    `Using VECTOR Data
+    <https://python-oracledb.readthedocs.io/en/latest/user_guide/vector_data_type.html>`_ - in the Oracle documentation
```
OK so we're going to point to the oracledb driver docs?   OK.   I will qualify the description here

suraj-ora-2020 replied:
Is there anything I need to do regarding this?

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5838 has been **merged**. Congratulations! :)

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5844 has been **merged**. Congratulations! :)
